### PR TITLE
For #2159 - new precision fields

### PR DESCRIPTION
**Pull request by TeranIvy**
Added two new field types as described in #2159, together with the relevant precisions,  `r_bl` and `r_phys`. Updated documentation, examples and tests. Also, many infrastructure file changes stem from updating the existing field infrastructure files.

### CODE REVIEW

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/operator/r_solver_operator_mod.F90
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/infrastructure/operator/r_solver_operator_mod.F90:1]: I can understand a pre-processed file being empty but why do we have some empty .F90 files?

File: src/psyclone/tests/test_files/dynamo0p3/mixed_sci_kernel_mod.f90 [hunk 1]
```diff
@@ -0,0 +1,116 @@
+! -----------------------------------------------------------------------------
```
Comment (arporter) [inline src/psyclone/tests/test_files/dynamo0p3/mixed_sci_kernel_mod.f90:1]: It's not obvious to me what this kernel adds to what we already have in `mixed_kernel_mod.f90`. Please could you explain? (Despite its name, it still boils down to a choice between 32- and 64-bit implementations.)

**Review by arporter — CHANGES_REQUESTED** [review]
Good work Iva.
All tests and examples are fine with compilation.
Updated UG builds fine.
I have some questions about some empty files and the new `mixed_sci_kernel_mod.f90` but apart from that it's tiny things.

File: src/psyclone/tests/test_files/dynamo0p3/15.1.3_aX_plus_Y_builtin.f90 [hunk 2]
```diff
@@ -31,22 +31,22 @@
...
+! Author: A. R. Porter, STFC Daresbury Lab
+! Modified: I. Kavcic, Met Office
 
 program single_invoke
 
   ! Description: single point-wise operation (Z = aX + Y)
   ! specified in an invoke call.
-  use constants_mod, only: r_def
-  use field_mod,     only: field_type
+  use constants_mod,  only: r_bl
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/15.1.3_aX_plus_Y_builtin.f90:41]: Updated

File: src/psyclone/tests/domain/lfric/lfric_builtins_test.py [hunk 3]
```diff
@@ -1030,14 +1030,21 @@ def test_aX_plus_Y(tmpdir, monkeypatch, annexed, dist_mem, fortran_writer):
 
     assert LFRicBuild(tmpdir).code_compiles(psy)
 
+    # Check for the correct 'use' module statements
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/lfric_builtins_test.py:1033]: Added

File: src/psyclone/tests/domain/lfric/lfric_builtins_test.py [hunk 4]
```diff
@@ -2611,8 +2618,25 @@ def test_a_times_X(tmpdir, monkeypatch, annexed, dist_mem, fortran_writer):
...
+    output_mod = (
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/lfric_builtins_test.py:2622]: Added

File: src/psyclone/tests/test_files/dynamo0p3/15.4.1_a_times_X_builtin.f90 [hunk 5]
```diff
@@ -31,20 +31,22 @@
...
 ! -----------------------------------------------------------------------------
-! Author A. R. Porter STFC Daresbury Lab
-! Modified I. Kavcic Met Office
+! Author: A. R. Porter, STFC Daresbury Lab
+! Modified: I. Kavcic, Met Office
 
 program single_invoke
 
   ! Description: single point-wise operation (scale a field: Y = aX)
   ! specified in an invoke call.
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/15.4.1_a_times_X_builtin.f90:40]: Updated

File: src/psyclone/tests/test_files/dynamo0p3/26.6.4_mixed_precision_rbl_vector.f90 [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors: R. W. Ford, STFC Daresbury Lab,
+!          I. Kavcic, Met Office
+!
+! Example where the field is dereferenced from an 'r_bl_vector_type'
+! and therefore has no type information. The LFRic rules state that an
+! 'r_bl_vector_type' can only contain fields of type
+! 'r_bl_field_type' in LFRic code. This is checked at runtime by the
+! algorithm code using a 'select' statement.
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/26.6.4_mixed_precision_rbl_vector.f90:39]: Yes, the `select` process happens in LFRic algorithms. I added a clarification.

File: src/psyclone/tests/test_files/dynamo0p3/26.6.5_mixed_precision_rphys_vector.f90 [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors: R. W. Ford, STFC Daresbury Lab,
+!          I. Kavcic, Met Office
+!
+! Example where the field is dereferenced from an 'r_phys_vector_type'
+! and therefore has no type information. The LFRic rules state that an
+! 'r_phys_vector_type' can only contain fields of type
+! 'r_phys_field_type' in LFRic code. This is checked at runtime by the
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/26.6.5_mixed_precision_rphys_vector.f90:38]: Again, added a clarification.

File: src/psyclone/tests/test_files/dynamo0p3/26.6.5_mixed_precision_rphys_vector.f90 [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors: R. W. Ford, STFC Daresbury Lab,
+!          I. Kavcic, Met Office
+!
+! Example where the field is dereferenced from an 'r_phys_vector_type'
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/26.6.5_mixed_precision_rphys_vector.f90:35]: Done.

File: src/psyclone/tests/test_files/dynamo0p3/26.6.4_mixed_precision_rbl_vector.f90 [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+! DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+! SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+! OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Authors: R. W. Ford, STFC Daresbury Lab,
+!          I. Kavcic, Met Office
+!
+! Example where the field is dereferenced from an 'r_bl_vector_type'
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/26.6.4_mixed_precision_rbl_vector.f90:35]: Done. Also updated the descriptions of algorithms that the new algorithms were copied from.

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/operator/r_solver_operator_mod.F90
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/infrastructure/operator/r_solver_operator_mod.F90:1]: It's not empty, the permissions were changed from exec to read so it appears empty in Git.

File: src/psyclone/tests/test_files/dynamo0p3/mixed_sci_kernel_mod.f90 (hunk 1 above)
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/mixed_sci_kernel_mod.f90:1]: I needed a kernel to just use scalars and fields of `r_bl` and `r_phys` precisions with no operators as there are no plans to add operators with these precisions for now. I can remove the new kernel and just reuse `mixed_kernel_mod.f90` with existing operator types if you're happy with it.

File: src/psyclone/tests/test_files/dynamo0p3/26.6.4_mixed_precision_rbl_vector.f90 [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+module vector_type
+
+  use constants_mod,    only : r_def
+  use vector_mod,       only : abstract_vector_type
+  use field_vector_mod, only : field_vector_type
+  use field_mod,        only : field_type
+  use testkern_mod,     only : testkern_type
+
+  type :: some_type
+     type(r_bl_field_vector_type) :: vec_type(10)
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/test_files/dynamo0p3/26.6.4_mixed_precision_rbl_vector.f90:50]: See the comment on `r_phys_field_vector_type` above.

File: doc/user_guide/dynamo0p3.rst [hunk 11]
```diff
@@ -530,28 +538,70 @@ layer. The rules for whether PSyclone requires information for
...
-``r_tran_field_vector_type`` it will assume that the actual field
-being referenced is of type ``field_type``, ``r_solver_field_type``,
-or ``r_tran_field_type`` respectively.
+In addition to fields, LFRic supports an abstract vector type for fields,
+used in the LFRic solver API. Please note that these structures are
+different from the :ref:`field vector <lfric-field-vector>` implementation
+of field bundles in the PSyclone LFRic API interface.
+
+The LFRic abstract vector type has precision-specific implementations.
+If PSyclone finds such specifically declared field vector argument in the
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply doc/user_guide/dynamo0p3.rst:582]: Not really, it is about an argument. I slighlty rephrased it, but I welcome feedback from a native speaker :smile:

File: src/psyclone/tests/domain/lfric/kern_call_arg_list_test.py [hunk 12]
```diff
@@ -436,6 +436,14 @@ def test_kerncallarglist_mixed_precision():
     assert create_arg_list.psyir_arglist[4].datatype.precision.name == "i_def"
     assert create_arg_list.psyir_arglist[5].datatype.precision.name == "r_tran"
 
+    create_arg_list = KernCallArgList(schedule.kernels()[3])
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/domain/lfric/kern_call_arg_list_test.py:439]: Done

File: src/psyclone/tests/dynkern_test.py [hunk 13]
```diff
@@ -190,6 +190,15 @@ def test_get_kernel_schedule_mixed_precision():
...
         assert sched.name == f"mixed_code_{8*precision}"
+    # The fourth kernel in 26.8 contains examples of precisions
+    # used in LFRic science.
+    kernel_precisions = ["r_bl", "r_phys"]
+    # Get the precision (in bytes) for each of these.
+    precisions = [LFRicConstants.PRECISION_MAP[name] for
+                  name in kernel_precisions]
+    sched = kernels[3].get_kernel_schedule()
+    for precision in precisions:
+        assert sched.name == f"mixed_sci_code_{8*precision}"
```
[the remark this answers is not in the document]
Comment (TeranIvy) [reply src/psyclone/tests/dynkern_test.py:201]: I updated the test example to reuse the existing `mixed_kernel_mod` and updated this test accordingly.

**Comment by TeranIvy** [comment]
Addressed review comments and tidied up field vector test examples. Ready for another look from @arporter.

File: doc/user_guide/dynamo0p3.rst [hunk 14]
```diff
@@ -530,28 +538,70 @@ layer. The rules for whether PSyclone requires information for
...
 +++++++++++++
 
-If PSyclone finds an argument that is declared as a
-``field_vector_type``, ``r_solver_field_vector_type`` or
-``r_tran_field_vector_type`` it will assume that the actual field
-being referenced is of type ``field_type``, ``r_solver_field_type``,
-or ``r_tran_field_type`` respectively.
+In addition to fields, LFRic supports an abstract vector type for fields,
+used in the LFRic solver API. Please note that these structures are
+different from the :ref:`field vector <lfric-field-vector>` implementation
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst:578]: I think we leave it for now and let LFRicians decide... :-)

File: doc/user_guide/dynamo0p3.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst]: Aha, the "a" was what was missing, thanks.

File: src/psyclone/tests/test_files/dynamo0p3/26.6.5_mixed_precision_rphys_vector.f90 [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+module vector_type
+
+  use constants_mod,    only : r_def
+  use vector_mod,       only : abstract_vector_type
+  use field_vector_mod, only : field_vector_type
+  use field_mod,        only : field_type
+  use testkern_mod,     only : testkern_type
+
+  type :: some_type
+     type(r_phys_field_vector_type) :: vec_type(10)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/26.6.5_mixed_precision_rphys_vector.f90:50]: I meant there was no use statement bringing it in, not that there was no object of that type. Anyway, I see you've fixed it so thank you :-)

File: src/psyclone/tests/test_files/dynamo0p3/infrastructure/operator/r_solver_operator_mod.F90
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/infrastructure/operator/r_solver_operator_mod.F90:1]: Ah! I should have checked the files themselves - apologies.

File: src/psyclone/tests/test_files/dynamo0p3/mixed_sci_kernel_mod.f90 (hunk 1 above)
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/mixed_sci_kernel_mod.f90]: That's great, thanks.

**Review by arporter — APPROVED** [review]
All good now:
 * Tests and examples all OK with compilation.
 * Ref. and User Guides build without problems.
 * All requested changes have been made.

Will proceed to merge.

**Pull request merged**